### PR TITLE
CASSANDRA-18802: Parallelized UCS compactions

### PR DESCRIPTION
**Pull request by blambov**
[CASSANDRA-18802](https://issues.apache.org/jira/browse/CASSANDRA-18802)

Implements parallelization of UCS compactions that would split the output into multiple sstables starting at known in advance boundaries. To do this, introduces a `CompositeLifecycleTransaction` that consists of multiple `PartialLifecycleTransaction` parts and only commits after all individual parts have completed, and creates multiple compaction tasks under the same composite transaction. Each individual task has a separate operation UUID, but to make it possible to recognize composite tasks their UUIDs have a part index as their sequence component (visible as "800n" as the second-to-last component of the UUID string), while non-parallelized ones have 0.

Major compaction is also changed to take advantage of parallelization. To make sure that it can be carried out on an active node, the number of threads used by a major compaction can be limited, to half the compaction threads by default. This is implemented by creating `CompositeCompactionTasks` that execute multiple tasks serially and grouping the major compaction tasks into a limited number of composites.

Because we do not currently support arbitrary filtering of the ranges of an sstable, parallelized compactions cannot use early open. Despite this, they are able to achieve comparable or better performance.

The first two commits bring in CASSANDRA-20092, which is needed for correct calculation of total compaction sizes. The next commit introduces some utilities that are helpful but not ultimately necessary for this patch (it can be easily adjusted to not use them; it is likely that SAI changes will bring these in independently). The final commit simplifies the method of creating the compaction-strategy-specific scanner list variation and is also optional.

### CODE REVIEW

File: src/java/org/apache/cassandra/db/ColumnFamilyStoreMBean.java [hunk 1]
```diff
@@ -52,6 +52,13 @@ public interface ColumnFamilyStoreMBean
      */
     public void forceMajorCompaction(boolean splitOutput) throws ExecutionException, InterruptedException;
 
+    /**
+     * force a major compaction of this column family
+     *
+     * @param permittedParallelism the maximum number of compaction threads that can be used by the operation
```
Comment (pcmanus) [inline src/java/org/apache/cassandra/db/ColumnFamilyStoreMBean.java]: Nit: wouldn't hurt to say that `permittedParallelism <= 0` is supported and treated specially.

File: src/java/org/apache/cassandra/db/compaction/AbstractCompactionStrategy.java [hunk 2]
```diff
@@ -331,56 +309,6 @@ public void metadataChanged(StatsMetadata oldMetadata, SSTableReader sstable)
...
-            return compressed / uncompressed;
-        }
-
-        public void close()
-        {
-            ISSTableScanner.closeAllAndPropagate(scanners, null);
-        }
-    }
-
     public ScannerList getScanners(Collection<SSTableReader> toCompact)
```
Comment (pcmanus) [inline src/java/org/apache/cassandra/db/compaction/AbstractCompactionStrategy.java]: This method can go away; it does the same thing than the default implementation of the interface.

File: src/java/org/apache/cassandra/db/compaction/AbstractCompactionStrategy.java [hunk 3]
```diff
@@ -62,7 +60,7 @@
  *    i/o done by compaction, and merging done at read time.
  *  - perform a full (maximum possible) compaction if requested by the user
  */
-public abstract class AbstractCompactionStrategy
+public abstract class AbstractCompactionStrategy implements ScannerFactory
```
Comment (pcmanus) [inline src/java/org/apache/cassandra/db/compaction/AbstractCompactionStrategy.java]: Nit: I find it slightly convoluted to implement `ScannerFactory`, and only a handful  of calls to `getScanner` on the strategy remains. I'd have added a protected method instead (for LCS to override)
```java
protected ScannerFactory scannerFactory()
{
   return ScannerFactory.DEFAULT;
}
```

File: src/java/org/apache/cassandra/db/compaction/CompactionController.java [hunk 4]
```diff
@@ -145,20 +151,20 @@ public Set<SSTableReader> getFullyExpiredSSTables()
      *
      * @param cfStore
      * @param compacting we take the drop-candidates from this set, it is usually the sstables included in the compaction
-     * @param overlapping the sstables that overlap the ones in compacting.
+     * @param overlappingSupplier function used to get the sstables that overlap the ones in compacting.
```
Comment (pcmanus) [inline src/java/org/apache/cassandra/db/compaction/CompactionController.java]: Nit: I'd rephrase to something like "called on the `compacting` sstables to compute the set of sstables that overlap with them if needed" (both to specify what the function argument is, and suggest that the reason this is a method is to avoid the computation if it's not needed (which I assume the reason?)).

File: src/java/org/apache/cassandra/db/compaction/CompactionManager.java [hunk 5]
```diff
@@ -376,24 +386,129 @@ public void run()
...
-                else
+                else if (tasks.size() == 1)
                 {
-                    task.execute(active);
+                    // If just one task, run it directly on this thread
+                    for (AbstractCompactionTask task : tasks)
+                        task.execute(active);
                     ranCompaction = true;
                 }
+                else
```
Comment (pcmanus) [inline src/java/org/apache/cassandra/db/compaction/CompactionManager.java:402]: Nit: I personally really find `if` that mix branches with and without brackets distracting/bad (and I've not seen others use them, though admittedly my personal experience with this particular code base is outdated of late). I could have swear this was explicitly prohibited by the coding style, but it's admittedly not clearly stated in the current form, and so be it. But this one trips me every time, so I figured I'd at least voice my opinion once, just so I feel better having voiced it, but it's quite a personal preference and feel free to ignore (and I won't bring it again).

File: src/java/org/apache/cassandra/db/compaction/CompactionManager.java [hunk 6]
```diff
@@ -1805,13 +1930,12 @@ public void obsoleteOriginals() {}
...
         }
 
-        CompactionStrategyManager strategy = cfs.getCompactionStrategyManager();
         try (SharedTxn sharedTxn = new SharedTxn(txn);
              SSTableRewriter fullWriter = SSTableRewriter.constructWithoutEarlyOpening(sharedTxn, false, groupMaxDataAge);
              SSTableRewriter transWriter = SSTableRewriter.constructWithoutEarlyOpening(sharedTxn, false, groupMaxDataAge);
              SSTableRewriter unrepairedWriter = SSTableRewriter.constructWithoutEarlyOpening(sharedTxn, false, groupMaxDataAge);
 
-             AbstractCompactionStrategy.ScannerList scanners = strategy.getScanners(txn.originals());
+             ScannerList scanners = ScannerFactory.DEFAULT.getScanners(txn.originals());
```
Comment (pcmanus) [inline src/java/org/apache/cassandra/db/compaction/CompactionManager.java]: This appears to disable the LCS "optimisation" with scanners. Am I misreading it? Is it intentional?

File: src/java/org/apache/cassandra/db/compaction/CompactionTask.java [hunk 7]
```diff
@@ -62,15 +64,17 @@ public class CompactionTask extends AbstractCompactionTask
...
 
-    public CompactionTask(ColumnFamilyStore cfs, LifecycleTransaction txn, long gcBefore)
+    public CompactionTask(ColumnFamilyStore cfs, AbstractCompactionStrategy strategy, ILifecycleTransaction txn, long gcBefore)
     {
-        this(cfs, txn, gcBefore, false);
+        this(cfs, strategy, txn, gcBefore, false);
     }
 
-    public CompactionTask(ColumnFamilyStore cfs, LifecycleTransaction txn, long gcBefore, boolean keepOriginals)
+    public CompactionTask(ColumnFamilyStore cfs, AbstractCompactionStrategy strategy, ILifecycleTransaction txn, long gcBefore, boolean keepOriginals)
```
Comment (pcmanus) [inline src/java/org/apache/cassandra/db/compaction/CompactionTask.java]: Nit: I've have pass a `ScannerFactory` type here instead of the strategy. That's the only part used and so makes the intent imo clearer. At which point I'd refuse `null` and manually pass `ScannerFactory.DEFAULT` when appropriate: that's one less thing that require readers to dig for that a `null` mean in this context.

File: src/java/org/apache/cassandra/db/compaction/CompactionTask.java [hunk 8]
```diff
@@ -84,68 +88,107 @@ protected int executeInternal(ActiveCompactionsTracker activeCompactions)
...
+            {
+                logger.debug("Compaction {} dropping expired sstables: {}", transaction.opIdString(), fullyExpiredSSTables);
+                actuallyCompact.removeAll(fullyExpiredSSTables);
+            }
 
             TimeUUID taskId = transaction.opId();
             // select SSTables to compact based on available disk space.
-            if (!buildCompactionCandidatesForAvailableDiskSpace(fullyExpiredSSTables, taskId))
+            final boolean hasExpirations = !fullyExpiredSSTables.isEmpty();
+            if (shouldReduceScopeForSpace() && !buildCompactionCandidatesForAvailableDiskSpace(actuallyCompact, hasExpirations, taskId)
```
Comment (pcmanus) [inline src/java/org/apache/cassandra/db/compaction/CompactionTask.java]: I'd add parenthesis to clarify the operator precedence we want (I suspect I'm not the only one that don't know operator precedence by heart; I reckon the line change here was probably meant to show the precedence, but I can't quite be sure it's not just your editor switching line automatically :)).

File: src/java/org/apache/cassandra/db/compaction/CompactionTask.java [hunk 9]
```diff
@@ -408,12 +450,8 @@ protected boolean buildCompactionCandidatesForAvailableDiskSpace(final Set<SSTab
                 // usually means we've run out of disk space
 
                 // but we can still compact expired SSTables
-                if(partialCompactionsAcceptable() && fullyExpiredSSTables.size() > 0 )
-                {
-                    // sanity check to make sure we compact only fully expired SSTables.
-                    assert transaction.originals().equals(fullyExpiredSSTables);
+                if (partialCompactionsAcceptable() && containsExpired)
```
Comment (pcmanus) [inline src/java/org/apache/cassandra/db/compaction/CompactionTask.java:459]: I'm not quite sure about this, but possibly because I don't fully understand this part.

My reading of the code before is that `reduceScope() == false` and `partialCompactionsAcceptable` implies that `transaction.originals() <= 1`, and so if we also have `fullyExpiredSSTable > 0`, then it did imply the one remaining sstable was expired, which the code asserted (it asserted we only had fully expired remaining, but I'm not there could have been more than one given the condition in `reduceScopeForLimitedSpace`).

With the change, `reduceScope() == false` and `partialCompactionsAcceptable` implies that `nonExpiredSSTables <= 1`. But afaict, even if we have expired sstables, that does not exclude having 1 remaining non expired sstable, which seems to contradict the "we can still compact expired SSTables".

Will all that said, I don't really understand what this "we can still compact expired SSTables" branch does in the first place, even before the changes. Because even if we exit through this branch, I don't see `CompactionTask#runMayThrow` doing anything with those "fully expired sstables" in the first place.

I'll lastly remark that `reduceCopeForLimitedSpace` is also overwritten by LCS, and I'm not sure how to reason about that one (neither before or after this patch).

File: src/java/org/apache/cassandra/db/compaction/CompositeCompactionTask.java [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.cassandra.utils.Throwables;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/// A composition of several compaction tasks into one. This is used to limit the parallelism of some compaction tasks
```
Comment (pcmanus) [inline src/java/org/apache/cassandra/db/compaction/CompositeCompactionTask.java]: Nit: maybe specify here that the composition executes task sequentially?

File: src/java/org/apache/cassandra/db/compaction/CompositeCompactionTask.java [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+
+    /// Add a task to the composition.
+    public CompositeCompactionTask addTask(AbstractCompactionTask task)
+    {
+        tasks.add(task);
+        return this;
+    }
+
+    @Override
+    protected int executeInternal(ActiveCompactionsTracker tracker)
```
Comment (pcmanus) [inline src/java/org/apache/cassandra/db/compaction/CompositeCompactionTask.java]: Nit: I couldn't find a single call of `execute` or `executeInternal` that was using the returned value. Maybe remove it and simplify instead of papering over it here?

File: src/java/org/apache/cassandra/db/compaction/LeveledCompactionTask.java [hunk 12]
```diff
@@ -100,6 +100,7 @@ public boolean reduceScopeForLimitedSpace(Set<SSTableReader> nonExpiredSSTables,
                             largestL0SSTable.onDiskLength(),
                             transaction.opId());
                 transaction.cancel(largestL0SSTable);
+                nonExpiredSSTables.remove(largestL0SSTable);
```
Comment (pcmanus) [inline src/java/org/apache/cassandra/db/compaction/LeveledCompactionTask.java:103]: Should the test at the beginning of this method be updated to check `nonExpiredSSTables` like the generic version of this in `CompactionTask`?

File: src/java/org/apache/cassandra/db/compaction/PendingRepairManager.java [hunk 13]
```diff
@@ -352,22 +346,25 @@ synchronized int getNumPendingRepairFinishedTasks()
         return count;
     }
 
-    synchronized AbstractCompactionTask getNextRepairFinishedTask()
+    synchronized Collection<AbstractCompactionTask> getNextRepairFinishedTasks()
     {
+        List<AbstractCompactionTask> tasks = null;
```
Comment (pcmanus) [inline src/java/org/apache/cassandra/db/compaction/PendingRepairManager.java]: Any specific reason to use `null` here instead of returning an empty collection? Using an empty collection would seem more consistent with  the other methods that are passed to `TaskSupplier` (and would even allow to expect `TaskSupplier.get()` to never return `null`).

File: src/java/org/apache/cassandra/db/compaction/AbstractCompactionStrategy.java [hunk 14]
```diff
@@ -180,17 +178,17 @@ public void shutdown()
      *
      * Is responsible for marking its sstables as compaction-pending.
      */
-    public abstract AbstractCompactionTask getNextBackgroundTask(final long gcBefore);
+    public abstract Collection<AbstractCompactionTask> getNextBackgroundTasks(final long gcBefore);
```
Comment (pcmanus) [inline src/java/org/apache/cassandra/db/compaction/AbstractCompactionStrategy.java:183]: Nit: the javadoc needs edits (`null` is never returned anymore in particular).

File: src/java/org/apache/cassandra/db/compaction/ShardManagerTrivial.java [hunk 15]
```diff
@@ -54,14 +57,21 @@ public double rangeSpanned(SSTableReader rdr)
...
-    public double calculateCombinedDensity(Set<? extends SSTableReader> sstables)
+    public double calculateCombinedDensity(Collection<SSTableReader> sstables)
     {
         double totalSize = 0;
         for (SSTableReader sstable : sstables)
             totalSize += sstable.onDiskLength();
         return totalSize;
     }
 
+    public <T> List<T> splitSSTablesInShards(Collection<SSTableReader> sstables,
```
Comment (pcmanus) [inline src/java/org/apache/cassandra/db/compaction/ShardManagerTrivial.java:69]: Nit: add `@Override`?

File: src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.java [hunk 16]
```diff
@@ -139,47 +139,80 @@ private static int atLeast2(int value, String str)
...
     {
         maybeUpdateShardManager();
         // The tasks are split by repair status and disk, as well as in non-overlapping sections to enable some
         // parallelism (to the amount that L0 sstables are split, i.e. at least base_shard_count). The result will be
         // split across shards according to its density. Depending on the parallelism, the operation may require up to
         // 100% extra space to complete.
         List<AbstractCompactionTask> tasks = new ArrayList<>();
-        List<Set<SSTableReader>> nonOverlapping = splitInNonOverlappingSets(filterSuspectSSTables(getSSTables()));
-        for (Set<SSTableReader> set : nonOverlapping)
+        if (permittedParallelism <= 0)
```
Comment (pcmanus) [inline src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.java]: Is this ever used? `CompactionManager.submitMaximal`, which I believe call this, already default it to `#compactors / 2` (it's also a tad confusing that that both places don't agree on what the default means).

File: src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.java [hunk 17]
```diff
@@ -139,47 +139,80 @@ private static int atLeast2(int value, String str)
...
         else if (w > 0)
-            return "T" + Integer.toString(w + 2);
+            return 'T' + Integer.toString(w + 2);
         else
             return "N";
     }
 
+    private TimeUUID nextTimeUUID()
+    {
+        return TimeUUID.Generator.nextTimeUUID().withSequence(0);
```
Comment (pcmanus) [inline src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.java:159]: It's not quite obvious to me why we're doing this. Can you explain/document?

File: src/java/org/apache/cassandra/db/compaction/unified/UnifiedCompactionTask.java [hunk 18]
```diff
@@ -36,26 +41,76 @@ public class UnifiedCompactionTask extends CompactionTask
...
+                                           shardManager.boundaries(numShards));
+    }
+
+    @Override
+    protected Range<Token> tokenRange()
+    {
+        return operationRange;
+    }
+
+    @Override
```
Comment (pcmanus) [inline src/java/org/apache/cassandra/db/compaction/unified/UnifiedCompactionTask.java:106]: Can you document why we should not reduce scope when `tokenRange() == null`?

File: src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.java (hunk 17 above)
Comment (pcmanus) [reply src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.java:159]: Reading further, I realize this is due to how `CompositeLifecycleTransaction` name its subtasks. It's fine, but again, worth documenting since the underlying reason is a bit "distant".

File: src/java/org/apache/cassandra/db/lifecycle/CompositeLifecycleTransaction.java [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+        this.obsoleteOriginalsRequested = false;
+    }
+
+    /// Register one part of the composite transaction. Every part must register itself before the composite transaction
+    /// is initialized and the parts are allowed to proceed.
+    /// @param part the part to register
+    public TimeUUID register(PartialLifecycleTransaction part)
+    {
+        int index = partsToCommitOrAbort.incrementAndGet();
+        return mainTransaction.opId().withSequence(index);
```
Comment (pcmanus) [inline src/java/org/apache/cassandra/db/lifecycle/CompositeLifecycleTransaction.java:79]: Nit: in practice, the idea of having linked operation ID between the parent and child tasks make sense. But this does kind of assume that the `mainTransaction` ID was generated with a 0 sequence (nothing would strongly break if that's not the case, but it's still somewhat assumed here), and that's not documented/easy to miss.

File: src/java/org/apache/cassandra/db/lifecycle/CompositeLifecycleTransaction.java [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+    }
+
+    /// Complete the initialization of the composite transaction. This must be called before any of the parts are
+    /// executed.
+    public void completeInitialization()
+    {
+        partsCount = partsToCommitOrAbort.get();
+        initializationComplete = true;
+        // TODO: Switch to trace before merging.
```
Comment (pcmanus) [inline src/java/org/apache/cassandra/db/lifecycle/CompositeLifecycleTransaction.java]: Flagging this so it doesn't get forgotten.

File: src/java/org/apache/cassandra/db/repair/CassandraValidationIterator.java [hunk 21]
```diff
@@ -221,7 +222,7 @@ public CassandraValidationIterator(ColumnFamilyStore cfs, SharedContext ctx, Col
 
         long gcBefore = dontPurgeTombstones ? Long.MIN_VALUE : getDefaultGcBefore(cfs, nowInSec);
         controller = new ValidationCompactionController(cfs, gcBefore);
-        scanners = cfs.getCompactionStrategyManager().getScanners(sstables, ranges);
+        scanners = ScannerFactory.DEFAULT.getScanners(sstables, ranges);
```
Comment (pcmanus) [inline src/java/org/apache/cassandra/db/repair/CassandraValidationIterator.java]: Not super fussed about it, but isn't it a slight regression that this is not allowed to rely on the compaction strategy to optimize the scanners?

File: src/java/org/apache/cassandra/tools/nodetool/Compact.java [hunk 22]
```diff
@@ -50,6 +50,12 @@ public class Compact extends NodeToolCmd
     @Option(title = "partition_key", name = {"--partition"}, description = "String representation of the partition key")
     private String partitionKey = EMPTY;
 
+    @Option(title = "jobs",
+            name = {"-j", "--jobs"},
+            description = "Use -j to specify the maximum number of threads to use for parallel compaction. " +
+                          "If not set, up to half the compaction threads will be used. " +
+                          "If set to 0, the major compaction will use all threads and will not permit other compactions to run until it completes (use with caution).")
```
Comment (pcmanus) [inline src/java/org/apache/cassandra/tools/nodetool/Compact.java:57]: I don't think this comment is up-to-date. Afaict, unset and 0 behave the same way.

File: src/java/org/apache/cassandra/utils/TimeUUID.java [hunk 23]
```diff
@@ -293,6 +293,22 @@ public static String toString(TimeUUID ballot, String kind)
         return ballot == null ? "null" : String.format("%s(%d:%s)", kind, ballot.uuidTimestamp(), ballot);
     }
 
+    public int sequence()
+    {
+        return (int) ((lsb >> 48) & 0x0000000000003FFFL);
+    }
+
+    /**
+     * Returns a new TimeUUID with the same timestamp as this one, but with the provided sequence value.
```
Comment (pcmanus) [inline src/java/org/apache/cassandra/utils/TimeUUID.java]: Let's call out that this method should be used with care as it removes "uniqueness" guarantees of the returned `TimeUUID` (same as we warn on `minAtUnixMillis` for instance).

File: src/java/org/apache/cassandra/db/ColumnFamilyStoreMBean.java (hunk 1 above)
Comment (blambov) [reply src/java/org/apache/cassandra/db/ColumnFamilyStoreMBean.java]: Good call. Verifying that it is indeed applied as described uncovered problems.

In fact, the strategy itself is the wrong place to apply this limit, because we may have other tasks (e.g. from other repair state or other disks).

Moved all the limit application to `CompactionStrategyManager`

File: src/java/org/apache/cassandra/db/compaction/AbstractCompactionStrategy.java (hunk 2 above)
Comment (blambov) [reply src/java/org/apache/cassandra/db/compaction/AbstractCompactionStrategy.java]: Done

File: src/java/org/apache/cassandra/db/compaction/AbstractCompactionStrategy.java (hunk 3 above)
Comment (blambov) [reply src/java/org/apache/cassandra/db/compaction/AbstractCompactionStrategy.java]: Done

File: src/java/org/apache/cassandra/db/compaction/AbstractCompactionStrategy.java (hunk 14 above)
Comment (blambov) [reply src/java/org/apache/cassandra/db/compaction/AbstractCompactionStrategy.java:183]: Changed, also in the next method.

File: src/java/org/apache/cassandra/db/compaction/CompactionController.java (hunk 4 above)
Comment (blambov) [reply src/java/org/apache/cassandra/db/compaction/CompactionController.java]: Done

File: src/java/org/apache/cassandra/db/compaction/CompactionManager.java (hunk 5 above)
Comment (blambov) [reply src/java/org/apache/cassandra/db/compaction/CompactionManager.java:402]: Changed

File: src/java/org/apache/cassandra/db/compaction/CompactionManager.java (hunk 6 above)
Comment (blambov) [reply src/java/org/apache/cassandra/db/compaction/CompactionManager.java]: Yes, it does. Taking advantage of it in places like this is the reason for the complexity I was trying to get rid of (the sstable set must be split for the individual strategy instances, then combined back; this adds cost to normal compactions).

Note that since CASSANDRA-8915, the cost of not using the LCS-specific scanner fell dramatically to just one key comparison per partition (i.e. one long comparison in most cases).

File: src/java/org/apache/cassandra/db/compaction/CompactionTask.java (hunk 7 above)
Comment (blambov) [reply src/java/org/apache/cassandra/db/compaction/CompactionTask.java]: Done; it was a little annoying, though, there were quite a few calls that needed to be adjusted.

I am tempted to completely remove the LCS-specific scanner.

File: src/java/org/apache/cassandra/db/compaction/CompactionTask.java (hunk 8 above)
Comment (blambov) [reply src/java/org/apache/cassandra/db/compaction/CompactionTask.java]: Added

File: src/java/org/apache/cassandra/db/compaction/CompactionTask.java (hunk 9 above)
Comment (blambov) [reply src/java/org/apache/cassandra/db/compaction/CompactionTask.java:459]: You are right, the new code would unnecessarily compact one sstable, while the old one would remove some of the expired ones.

Changed this part to remove any remaining nonExpired sstables from the transaction, which should result in leaving only all of expired ones (which is an improvement over the old behaviour). The effect of not having anything to compact is to complete the transaction after doing nothing, which in turn will delete the expired sstables.

File: src/java/org/apache/cassandra/db/compaction/CompositeCompactionTask.java (hunk 10 above)
Comment (blambov) [reply src/java/org/apache/cassandra/db/compaction/CompositeCompactionTask.java]: Done

File: src/java/org/apache/cassandra/db/compaction/CompositeCompactionTask.java (hunk 11 above)
Comment (blambov) [reply src/java/org/apache/cassandra/db/compaction/CompositeCompactionTask.java]: Done

File: src/java/org/apache/cassandra/db/compaction/LeveledCompactionTask.java (hunk 12 above)
Comment (blambov) [reply src/java/org/apache/cassandra/db/compaction/LeveledCompactionTask.java:103]: Yes, done

File: src/java/org/apache/cassandra/db/compaction/PendingRepairManager.java (hunk 13 above)
Comment (blambov) [reply src/java/org/apache/cassandra/db/compaction/PendingRepairManager.java]: Done

File: src/java/org/apache/cassandra/db/compaction/ShardManagerTrivial.java (hunk 15 above)
Comment (blambov) [reply src/java/org/apache/cassandra/db/compaction/ShardManagerTrivial.java:69]: Done

File: src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.java (hunk 17 above)
Comment (blambov) [reply src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.java:159]: Added

File: src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.java (hunk 16 above)
Comment (blambov) [reply src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.java]: No, it was not (apart from a test), and has been removed in the current code.

File: src/java/org/apache/cassandra/db/compaction/unified/UnifiedCompactionTask.java (hunk 18 above)
Comment (blambov) [reply src/java/org/apache/cassandra/db/compaction/unified/UnifiedCompactionTask.java:106]: Added

File: src/java/org/apache/cassandra/db/lifecycle/CompositeLifecycleTransaction.java (hunk 19 above)
Comment (blambov) [reply src/java/org/apache/cassandra/db/lifecycle/CompositeLifecycleTransaction.java:79]: Added comment to the constructor.

File: src/java/org/apache/cassandra/db/lifecycle/CompositeLifecycleTransaction.java (hunk 20 above)
Comment (blambov) [reply src/java/org/apache/cassandra/db/lifecycle/CompositeLifecycleTransaction.java]: Changed

File: src/java/org/apache/cassandra/db/repair/CassandraValidationIterator.java (hunk 21 above)
Comment (blambov) [reply src/java/org/apache/cassandra/db/repair/CassandraValidationIterator.java]: This is correct, that optimization will no longer apply here.

We have three options on this:
- Leave as is (special scanner applies only to compactions, no cost to get list)
- Roll back https://github.com/apache/cassandra/pull/3688/commits/b24959861d38b139720b36791714047261efcf1d and associated follow-ups (special scanner applies everywhere, all operations pay a collection cost when getting the scanner list)
- Drop the special LCS scanner altogether (LCS pays a cost of up to 1 extra key comparison per partition in exchange for lower call polymorphism).

If you are not happy with the first option, I can run some benchmarks to see if the special scanner actually saves anything to choose between the other two.

File: src/java/org/apache/cassandra/tools/nodetool/Compact.java (hunk 22 above)
Comment (blambov) [reply src/java/org/apache/cassandra/tools/nodetool/Compact.java:57]: This is now fixed,

File: src/java/org/apache/cassandra/utils/TimeUUID.java (hunk 23 above)
Comment (blambov) [reply src/java/org/apache/cassandra/utils/TimeUUID.java]: Added

File: src/java/org/apache/cassandra/db/repair/CassandraValidationIterator.java (hunk 21 above)
Comment (pcmanus) [reply src/java/org/apache/cassandra/db/repair/CassandraValidationIterator.java]: In principle, I love simplifying code and would rather we get rid of the special case, especially since LCS is in my mind kind of deprecated (in favor of UCS). But I'm sure plenty of users still use LCS and the truth is that I have no good intuition on the impact of this. What's the concrete impact of removing that LCS scanner for compaction? and for validation/anti-compaction? I feel the right decision mostly depend on the answer to those questions (unless we've officially deprecated LCS, but I don't think that's the case)?

I do understand doing benchmark might be time consuming though. I just feel I'm lacking proper data to have an opinion at the moment. But my only concern is making sure LCS users don't experience a clearly noticeable performance because of this; I'm not attached at all to that special scanner otherwise.

Comment (blambov) [reply src/java/org/apache/cassandra/db/repair/CassandraValidationIterator.java]: Reverted the change, opening a separate ticket to deal with the `ScannerList` complexity.

Comment (blambov) [reply src/java/org/apache/cassandra/db/repair/CassandraValidationIterator.java]: CASSANDRA-20134

**Review by pcmanus — APPROVED** [review]
Lgtm, awesome work.

**Comment by blambov** [comment]
Committed as https://github.com/apache/cassandra/commit/54e46880690bd5effb31116986292c1bdc9e891e.
